src/mbgl/renderer/layers/render_circle_layer.cpp

```diff
@@ -236,4 +262,162 @@ bool RenderCircleLayer::queryIntersectsFeature(const GeometryCoordinates& queryG
     return false;
 }
 
+#if MLN_DRAWABLE_RENDERER
+namespace {
+
+struct alignas(16) CircleInterpolateUBO {
+    float color_t;
+    float radius_t;
+    float blur_t;
+    float opacity_t;
+    float stroke_color_t;
+    float stroke_width_t;
+    float stroke_opacity_t;
+    float padding;
+};
+static_assert(sizeof(CircleInterpolateUBO) % 16 == 0);
+
+constexpr auto CircleShaderGroupName = "CircleShader";
+constexpr auto CircleInterpolateUBOName = "CircleInterpolateUBO";
+constexpr auto VertexAttribName = "a_pos";
+
+} // namespace
+
+void RenderCircleLayer::update(gfx::ShaderRegistry& shaders,
+                               gfx::Context& context,
+                               const TransformState& state,
+                               [[maybe_unused]] const RenderTree& renderTree,
+                               UniqueChangeRequestVec& changes) {
+    std::unique_lock<std::mutex> guard(mutex);
+
+    if (!renderTiles || renderTiles->empty()) {
+        removeAllDrawables();
+        return;
+    }
+
+    // Set up a layer group
+    if (!layerGroup) {
+        if (auto layerGroup_ = context.createTileLayerGroup(layerIndex, /*initialCapacity=*/64, getID())) {
+            layerGroup_->setLayerTweaker(std::make_shared<CircleLayerTweaker>(evaluatedProperties));
+            setLayerGroup(std::move(layerGroup_), changes);
+        }
+    }
+    auto* tileLayerGroup = static_cast<TileLayerGroup*>(layerGroup.get());
+
+    if (!circleShaderGroup) {
+        circleShaderGroup = shaders.getShaderGroup(CircleShaderGroupName);
+    }
+    if (!circleShaderGroup) {
+        removeAllDrawables();
+        return;
+    }
+
+    std::unique_ptr<gfx::DrawableBuilder> circleBuilder;
+    constexpr auto renderPass = RenderPass::Translucent;
+
+    if (!(mbgl::underlying_type(renderPass) & evaluatedProperties->renderPasses)) {
+        return;
+    }
+
+    stats.drawablesRemoved += tileLayerGroup->removeDrawablesIf(
+        [&](gfx::Drawable& drawable) { return drawable.getTileID() && !hasRenderTile(*drawable.getTileID()); });
+
+    const auto& evaluated = static_cast<const CircleLayerProperties&>(*evaluatedProperties).evaluated;
+
+    for (const RenderTile& tile : *renderTiles) {
+        const auto& tileID = tile.getOverscaledTileID();
+
+        const LayerRenderData* renderData = getRenderDataForPass(tile, renderPass);
+        if (!renderData || !renderData->bucket || !renderData->bucket->hasData()) {
+            removeTile(renderPass, tileID);
+            continue;
+        }
+
+        const auto& bucket = static_cast<const CircleBucket&>(*renderData->bucket);
+        const auto vertexCount = bucket.vertices.elements();
+        const auto& paintPropertyBinders = bucket.paintPropertyBinders.at(getID());
+
+        const auto prevBucketID = getRenderTileBucketID(tileID);
```

Comment:
During the meeting it was mentioned this could have a little more explanation (that is, why this is needed).

Reply:
This code, and the terse comment below, exists in each layer, so an explanatory comment probably belongs somewhere more central.  I'll add something to `getRenderTileBucketID` separately.